Mimir.MongoDB/Repositories/ProductRepository.cs

```diff
@@ -40,4 +55,124 @@ public async Task<ProductDocument> GetByProductIdAsync(Guid productId)
+        }
+
+        return filter;
+    }
+
+    private IExecutable<ProductDocument> ApplySorting(ProductFilter productFilter, FilterDefinition<ProductDocument> filter,
+        IFindFluent<ProductDocument, ProductDocument> find)
+    {
+        productFilter.SortDirection ??= Enums.SortDirection.Ascending;
+
+        switch (productFilter.SortBy)
+        {
+            case ProductSortBy.Cp:
+                break;
+            case ProductSortBy.Crystal:
+                break;
+            case ProductSortBy.Grade:
+                return SortByGrade(productFilter, find);
+            case ProductSortBy.Price:
+                return SortByPrice(filter, productFilter);
+            case ProductSortBy.UnitPrice:
+                return SortByUnitPrice(productFilter, filter, find);
+        }
+
+        return find.AsExecutable();
+    }
+
+    private static MongoDbFindFluentExecutable<ProductDocument> SortByGrade(ProductFilter productFilter,
+        IFindFluent<ProductDocument, ProductDocument> find)
+    {
+        var sortBuilder = Builders<ProductDocument>.Sort;
+        find = productFilter.SortDirection == Enums.SortDirection.Ascending
+            ? find.Sort(sortBuilder.Ascending("Object.TradableItem.Grade"))
+            : find.Sort(sortBuilder.Descending("Object.TradableItem.Grade"));
+
+        return find.AsExecutable();
+    }
+
+    private MongoDbAggregateFluentExecutable<ProductDocument> SortByPrice(FilterDefinition<ProductDocument> filter,
+        ProductFilter productFilter)
+    {
+        var convertStage = new BsonDocument("$addFields", new BsonDocument
+        {
+            { "convertedPrice", new BsonDocument("$toLong", "$Object.Price.MajorUnit") },
+        });
+        var sortStage = new BsonDocument("$sort", new BsonDocument("convertedPrice",
+            productFilter.SortDirection == Enums.SortDirection.Ascending ? 1 : -1));
+
+        return _collection.Aggregate()
+            .Match(filter)
+            .AppendStage<ProductDocument>(convertStage)
+            .AppendStage<ProductDocument>(sortStage)
+            .AsExecutable();
+    }
+
+    private IExecutable<ProductDocument> SortByUnitPrice(ProductFilter productFilter, FilterDefinition<ProductDocument> filter, IFindFluent<ProductDocument, ProductDocument> find)
+    {
+        // ProductDocument.UnitPrice is null for FungibleAssetValue
+        if (productFilter.ProductType == ProductType.FungibleAssetValue)
+        {
+            return SortFungibleAssetValueByUnitPrice(productFilter, filter);
+        }
+
+        find = productFilter.SortDirection == Enums.SortDirection.Ascending
+            ? find.SortBy(x => x.UnitPrice)
+            : find.SortByDescending(x => x.UnitPrice);
+
+        return find.AsExecutable();
+    }
+
+    private MongoDbAggregateFluentExecutable<ProductDocument> SortFungibleAssetValueByUnitPrice(ProductFilter productFilter,
+        FilterDefinition<ProductDocument> filter)
```

Comment:
How about just accept the `SortDirection` type instead of the `ProductFilter` type argument?